File: packages/mcp/src/parser/extractors/schematics.ts
import type { Ingredient, ItemForm, RawClass, Schematic, SchematicType } from '../types.js';
import { getNumber, getString, isRecord } from '../util.js';
import { extractClassNames, humaniseClassName } from '../normalise/classRef.js';
import { parseItemAmountList } from '../normalise/ingredients.js';
import { toDisplayAmount } from '../normalise/fluids.js';

/** Lookups built from already-parsed items, recipes and buildings. */
export interface SchematicLookups {
  itemForm: Map<string, ItemForm>;
  itemDisplay: Map<string, string>;
  itemClasses: Set<string>;
  productionRecipeClasses: Set<string>;
  /** Build-gun recipe class name → building class name it constructs. */
  buildRecipeToBuilding: Map<string, string>;
}

const TYPE_MAP: Record<string, SchematicType> = {
  EST_Milestone: 'milestone',
  EST_MAM: 'mam',
  EST_ResourceSink: 'awesome_shop',
  EST_HardDrive: 'hard_drive',
  EST_Alternate: 'hard_drive',
  EST_Tutorial: 'tutorial',
};

function mapType(raw: string): SchematicType {
  return TYPE_MAP[raw] ?? 'other';
}

/** Cost lines carry no production rate, so perMinute is 0. */
function mapCost(raw: string, lookups: SchematicLookups): Ingredient[] {
  return parseItemAmountList(raw).map((item): Ingredient => {
    const form = lookups.itemForm.get(item.className) ?? 'solid';
    const { amount, unit } = toDisplayAmount(item.amount, form);
    return {
      itemClassName: item.className,
      displayName: lookups.itemDisplay.get(item.className) ?? humaniseClassName(item.className),
      amount,
      perMinute: 0,
      unit,
    };
  });
}

function collectUnlockField(unlock: Record<string, unknown>, field: string): string[] {
  const value = unlock[field];
  return typeof value === 'string' ? extractClassNames(value) : [];
}

export function extractSchematic(raw: RawClass, lookups: SchematicLookups): Schematic {
  const className = getString(raw, 'ClassName');

  const unlocksRecipes: string[] = [];
  const unlocksBuildings: string[] = [];
  const unlocksItems: string[] = [];

  const rawUnlocks = raw['mUnlocks'];
  if (Array.isArray(rawUnlocks)) {
    for (const unlock of rawUnlocks) {
      if (!isRecord(unlock)) {
        continue;
      }
      // Recipes (and blueprint recipes) unlock either a production recipe or,
      // when the recipe is a build-gun recipe, a building.
      for (const recipeClass of [
        ...collectUnlockField(unlock, 'mRecipes'),
        ...collectUnlockField(unlock, 'mBlueprints'),
      ]) {
        if (lookups.productionRecipeClasses.has(recipeClass)) {
          unlocksRecipes.push(recipeClass);
        } else {
          const building = lookups.buildRecipeToBuilding.get(recipeClass);
          if (building !== undefined) {
            unlocksBuildings.push(building);
          }
        }
      }
      // Items granted or unlocked directly.
      for (const itemClass of [
        ...collectUnlockField(unlock, 'mItemDescriptors'),
        ...collectUnlockField(unlock, 'mItemsToGive'),
      ]) {
        if (lookups.itemClasses.has(itemClass)) {
          unlocksItems.push(itemClass);
        }
      }
    }
  }

  return {
    className,
    displayName: getString(raw, 'mDisplayName') || humaniseClassName(className),
    type: mapType(getString(raw, 'mType')),
    tier: getNumber(raw, 'mTechTier', 0),
    cost: mapCost(getString(raw, 'mCost'), lookups),
    unlocksRecipes: dedupe(unlocksRecipes),
    unlocksBuildings: dedupe(unlocksBuildings),
    unlocksItems: dedupe(unlocksItems),
  };
}

function dedupe(values: string[]): string[] {
  return [...new Set(values)];
}

File: packages/sf-game-data/extract/sf-game-data-parse/Schematics.cs
using System.Text.Json;

namespace SfGameData.Parse;

// Lookups built from already-parsed items, recipes and buildings.
public sealed class SchematicLookups
{
    public Dictionary<string, string> ItemForm { get; init; } = [];
    public Dictionary<string, string> ItemDisplay { get; init; } = [];
    public HashSet<string> ItemClasses { get; init; } = [];
    public HashSet<string> ProductionRecipeClasses { get; init; } = [];
    public Dictionary<string, string> BuildRecipeToBuilding { get; init; } = [];
}

// Ported from src/parser/extractors/schematics.ts.
public static class Schematics
{
    private static readonly Dictionary<string, string> TypeMap = new()
    {
        ["EST_Milestone"] = "milestone",
        ["EST_MAM"] = "mam",
        ["EST_ResourceSink"] = "awesome_shop",
        ["EST_HardDrive"] = "hard_drive",
        ["EST_Alternate"] = "hard_drive",
        ["EST_Tutorial"] = "tutorial",
    };

    private static string MapType(string raw) => TypeMap.GetValueOrDefault(raw, "other");

    // Cost lines carry no production rate, so perMinute is 0.
    private static List<Ingredient> MapCost(string raw, SchematicLookups lookups)
        => Normalise.ParseItemAmountList(raw).Select(item =>
        {
            var form = lookups.ItemForm.GetValueOrDefault(item.ClassName, "solid");
            var display = Normalise.ToDisplayAmount(item.Amount, form);
            return new Ingredient
            {
                ItemClassName = item.ClassName,
                DisplayName = lookups.ItemDisplay.GetValueOrDefault(item.ClassName, ""),
                Amount = display.Amount,
                PerMinute = 0,
                Unit = display.Unit,
            };
        }).ToList();

    // The TS reads unlock[field] only when it is a string; GetString returns "" for
    // a missing/non-string field, and ExtractClassNames("") is [], so this matches.
    private static List<string> CollectUnlockField(JsonElement unlock, string field)
        => ClassRef.ExtractClassNames(Util.GetString(unlock, field));

    public static Schematic ExtractSchematic(JsonElement raw, SchematicLookups lookups)
    {
        var className = Util.GetString(raw, "ClassName");

        var unlocksRecipes = new List<string>();
        var unlocksBuildings = new List<string>();
        var unlocksItems = new List<string>();

        var rawUnlocks = Util.GetElement(raw, "mUnlocks");
        if (rawUnlocks.ValueKind == JsonValueKind.Array)
        {
            foreach (var unlock in rawUnlocks.EnumerateArray())
            {
                if (!Util.IsRecord(unlock))
                {
                    continue;
                }
                foreach (var recipeClass in CollectUnlockField(unlock, "mRecipes")
                    .Concat(CollectUnlockField(unlock, "mBlueprints")))
                {
                    if (lookups.ProductionRecipeClasses.Contains(recipeClass))
                    {
                        unlocksRecipes.Add(recipeClass);
                    }
                    else if (lookups.BuildRecipeToBuilding.TryGetValue(recipeClass, out var building))
                    {
                        unlocksBuildings.Add(building);
                    }
                }
                foreach (var itemClass in CollectUnlockField(unlock, "mItemDescriptors")
                    .Concat(CollectUnlockField(unlock, "mItemsToGive")))
                {
                    if (lookups.ItemClasses.Contains(itemClass))
                    {
                        unlocksItems.Add(itemClass);
                    }
                }
            }
        }

        return new Schematic
        {
            ClassName = className,
            DisplayName = Util.GetString(raw, "mDisplayName"),
            Type = MapType(Util.GetString(raw, "mType")),
            Tier = Util.GetNumber(raw, "mTechTier", 0),
            Cost = MapCost(Util.GetString(raw, "mCost"), lookups),
            UnlocksRecipes = Dedupe(unlocksRecipes),
            UnlocksBuildings = Dedupe(unlocksBuildings),
            UnlocksItems = Dedupe(unlocksItems),
        };
    }

    private static List<string> Dedupe(List<string> values)
    {
        var seen = new HashSet<string>();
        var output = new List<string>();
        foreach (var value in values)
        {
            if (seen.Add(value))
            {
                output.Add(value);
            }
        }
        return output;
    }
}
